src/Security/PasswordExpirationMiddleware.php

```diff
@@ -0,0 +1,265 @@
+
+        if ($session && $session->get(static::SESSION_KEY_ALLOW_CURRENT_REQUEST)) {
+            // allow current request and skip the expiration check, but for only the current
+            // request, so we're deleting the flag from the session so it's not affecting other
+            // requests.
+            // This flag would usually be set from within $handler->authenticateRequest()
+            $session->clear(static::SESSION_KEY_ALLOW_CURRENT_REQUEST);
+
+            return null;
+        }
+
+        $user = Security::getCurrentUser();
+        if ($user && $user->isPasswordExpired()) {
+            if ($response = $this->handleExpiredPassword($request)) {
+                return $response;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Check if we have a redirect to a password change form registered
+     * and redirect there if possible.
+     * Otherwise, deauthenticate the user by resetting it for this request,
+     * since we should treat ones with expired passwords as unauthorised.
+     *
+     * @param HTTPRequest $request
+     *
+     * @return HTTPResponse|null
+     */
+    protected function handleExpiredPassword(HTTPRequest $request): ?HTTPResponse
+    {
+        $session = $request->getSession();
+
+        $sessionRedirectUrl = $session->get(static::SESSION_KEY_REDIRECT);
+        $defaultRedirectUrl = static::config()->get('default_redirect');
+
+        if ($sessionRedirectUrl || $defaultRedirectUrl) {
+            $redirectUrl = $this->absoluteUrl($sessionRedirectUrl ?? $defaultRedirectUrl);
+        } else {
+            $redirectUrl = null;
+        }
+
+        if (!$session || !$redirectUrl) {
+            Security::setCurrentUser(null);
+            return null;
+        }
+
+        $whitelist = [];
+
+        $currentUrl = $this->absoluteUrl($request->getURL(true));
+
+        $allowedUrls = array_merge([
+            $redirectUrl
+        ], array_map(
+            function ($url) {
+                return $this->absoluteUrl($url);
+            },
+            $whitelist
+        ));
+
+        if (in_array($currentUrl, $allowedUrls, true)) {
+            return null;
+        }
+
+        $allowedStartswith = static::config()->get('whitelisted_url_startswith');
+        if (is_array($allowedStartswith)) {
+            foreach ($allowedStartswith as $pattern) {
+                $startswith = $this->absoluteUrl($pattern);
+
+                if (strncmp($currentUrl, $startswith, strlen($startswith)) === 0) {
+                    return null;
+                }
+            }
+        }
+
+        return $this->redirectOrForbid($request, $redirectUrl);
+    }
+
+    /**
+     * Builds an absolute URL for the given path, adds base url
+     * if the path configured as absolute
+     *
+     * @param string $url
+     *
+     * @return string
+     */
+    protected static function absoluteUrl($url): string
+    {
+        if (substr($url, 0, 1) === '/') {
```

Comment:
Just checking, we are now saying that `//localhost/foo` is not absolute, right? This check would return true for that URL

Reply:
> Just checking, we are now saying that //localhost/foo is not absolute, right? This check would return true for that URL

Not until ss5 at the earliest.

Reply:
I've explained at https://github.com/silverstripe/silverstripe-framework/issues/9111#issuecomment-527009973 why you should not be treating root-relative urls as though they were base-relative. If you have a URL with a leading slash that you don't want to treat as a root-relative url, your data being put into the method is wrong.

Reply:
> Just checking, we are now saying that //localhost/foo is not absolute, right?

Technically, it's not an absolute URL, but it is still either an `absolute path` or a `network location` (relative URL starting with //). This check is basically for paths, not urls.

This is a good catch, by the way. Even though it's an edge case, in case of a network location this would produce incorrect result. I'll patch it up and cover with a test.